File: JavaToKotlin/app/src/main/java/io/dustin/apps/board/domain/qna/answer/service/ReadAnswerService.java
package io.dustin.apps.board.domain.qna.answer.service;

import io.dustin.apps.board.domain.qna.answer.model.Answer;
import io.dustin.apps.board.domain.qna.answer.repository.AnswerRepository;
import io.dustin.apps.common.exception.DataNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ReadAnswerService {

    private final AnswerRepository answerRepository;

    @Transactional(readOnly = true)
    public Answer findByQuestionId(Long questionId) {
        return answerRepository.findByQuestionId(questionId);
    }



    @Transactional(readOnly = true)
    public Answer findById(Long id) {
        Optional<Answer> optional = answerRepository.findById(id);
        if(optional.isPresent()) {
            return optional.get();
        } else {
            return null;
        }
    }

    @Transactional(readOnly = true)
    public Answer findByIdOrThrow(Long id) {
        Optional<Answer> optional = this.answerRepository.findById(id);
        if(optional.isPresent()) {
            return optional.get();
        } else {
            throw new DataNotFoundException("""
                    id [#1]로 조회된 게시물이 없습니다.""".replace("#1", String.valueOf(id)));
        }
    }

}

File: apps/src/main/kotlin/io/dustin/apps/board/domain/qna/answer/service/ReadAnswerService.kt
package io.dustin.apps.board.domain.qna.answer.service

import io.dustin.apps.board.domain.qna.answer.model.Answer
import io.dustin.apps.board.domain.qna.answer.repository.AnswerRepository
import io.dustin.apps.board.domain.qna.question.model.Question
import io.dustin.apps.common.exception.DataNotFoundException
import io.dustin.apps.common.repository.findByIdOrThrow
import io.dustin.apps.common.utils.notFoundEntity
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class ReadAnswerService(

    private val answerRepository: AnswerRepository

) {

    fun getAnswerObj(questionId: Long): Answer? {
        return answerRepository.findByQuestionId(questionId)
    }

    /**
     * questionId로 Answer 객체를 찾는다.
     */
    fun findByQuestionId(questionId: Long): Answer {
        val errorMessage = "질문객체 고유값[$questionId]로 조회된 정보가 없습니다."
        return answerRepository.findByQuestionId(questionId) ?: notFoundEntity(errorMessage)
    }

    fun findByIdOrNull(answerId: Long): Answer? {
        return answerRepository.findByIdOrNull(answerId)
    }

    fun findByIdOrThrow(answerId: Long): Answer {
        return answerRepository.findByIdOrThrow(answerId)
    }
}

